components/tidb_query_datatype/src/codec/mysql/json/path_expr.rs

```diff
@@ -73,82 +65,325 @@ impl PathExpression {
+        }
+    }
 
-    lazy_static::lazy_static! {
-        static ref RE: Regex = Regex::new(PATH_EXPR_LEG_RE_STR).unwrap();
+    // returns the current index on the slice
+    fn current_index(&mut self) -> usize {
+        match self.char_iterator.peek() {
+            Some((start, _)) => *start,
+            None => self.input.len(),
+        }
     }
-    let mut legs = vec![];
-    let mut flags = PathExpressionFlag::default();
-    let mut last_end = 0;
-    for m in RE.find_iter(expr) {
-        let (start, end) = (m.start(), m.end());
-        // Check all characters between two legs are blank.
-        if expr
-            .index(last_end..start)
-            .char_indices()
-            .any(|(_, c)| !c.is_ascii_whitespace())
-        {
-            return Err(box_err!("Invalid JSON path: {}", path_expr));
+
+    // trim_while_spaces remove spaces from the position
+    fn trim_white_spaces(&mut self) {
+        while self
+            .char_iterator
+            .next_if(|(_, ch)| ch.is_whitespace())
+            .is_some()
+        {}
+    }
+
+    // returns whether the input has reached the end
+    fn reached_end(&mut self) -> bool {
+        return self.char_iterator.peek().is_none();
+    }
+
+    fn next_key(&mut self) -> Result<PathExpressionToken> {
+        let (start, _) = self.char_iterator.next().unwrap();
+
+        self.trim_white_spaces();
+        if self.reached_end() {
+            return Err(box_json_path_err!(self.current_index()));
         }
-        last_end = end;
-
-        let next_char = expr.index(start..).chars().next().unwrap();
-        if next_char == '[' {
-            // The leg is an index of a JSON array.
-            let leg = expr[start + 1..end].trim();
-            let index_str = leg[0..leg.len() - 1].trim();
-            let index = if index_str == PATH_EXPR_ASTERISK {
-                flags |= PATH_EXPRESSION_CONTAINS_ASTERISK;
-                PATH_EXPR_ARRAY_INDEX_ASTERISK
-            } else {
-                box_try!(index_str.parse::<i32>())
-            };
-            legs.push(PathLeg::Index(index))
-        } else if next_char == '.' {
-            // The leg is a key of a JSON object.
-            let mut key = expr[start + 1..end].trim().to_owned();
-            if key == PATH_EXPR_ASTERISK {
-                flags |= PATH_EXPRESSION_CONTAINS_ASTERISK;
-            } else if key.starts_with('"') {
-                // We need to unquote the origin string.
-                key = unquote_string(&key[1..key.len() - 1])?;
+
+        match *self.char_iterator.peek().unwrap() {
+            (_, '*') => {
+                self.char_iterator.next().unwrap();
+
+                Ok(PathExpressionToken::Leg((
+                    PathLeg::Key(PATH_EXPR_ASTERISK.to_string()),
+                    Position {
+                        start,
+                        end: self.current_index(),
+                    },
+                )))
+            }
+            (mut key_start, '"') => {
```

Comment:
can you please add a test like `'$."a\t"'`. MySQL will report an invalid path since `"a\t"` is not a valid json string. 

it is a corner case, I think it is ok to add the check in the future. ignore this if the test case is passed

Reply:
Oops! I didn't realize it. I'll add a test case and fix.

Reply:
Fixed. I checked the ascii control character. However, after unquote, I cannot tell the concrete position of the original input, so I only return the beginning position of the string literal.